// react-native-roll-picker/lib/Picker.js
/**
 * Created by yjy on 16/8/3.
 *
 * props:
 *  data: 数据
 *  name: rowData[name] 返回列表数据
 *  onRowChange: (index) => {} 被选中的index
 *
 * method:
 *  setDataSource(data): 刷新数据
 *
 */

import React, { Component } from 'react';
import {
    Text,
    View,
    Image,
    TouchableOpacity,
    ScrollView,
    ListView
} from 'react-native';

import { findNodeHandle } from 'react-native';
import { NativeModules } from 'react-native';
const RCTUIManager =  NativeModules.UIManager;

import {CSS} from './CSS'

export default class Picker extends Component {
    constructor(props) {
        super(props);
        let data = props.data;
        this.onScrollCount = 0;
        this.state = {
            data: data
        };
    }

    setDataSource(data) {
        if(this.refs._ScrollView) {
            this.refs._ScrollView.scrollTo({y: 0, animated: false});
        }
        this.setState({data: data});
    }

    getItem(size) {
        if(this.state.data.length == 0) {
            return false;
        }
        let arr = this.state.data;
        const justifyStyle = this.props.pickerStr === 'hour' ? {justifyContent: 'flex-end'} : {justifyContent: 'flex-start', marginLeft: 20}
        return arr.map((item, i) => {
            return (
                <View key={i} style={[justifyStyle, {height: 40, alignItems: 'center', flexDirection: 'row'}]}>
                    <Text style={{fontSize: CSS.pixel(50), color: size == 'big' ? '#218FD8' : '#707070', backgroundColor: 'rgba(0,0,0,0)'}}>
                        {this.props.name ? item[this.props.name] : item}
                    </Text>
					<Text style={{width: 60, fontSize: 20, color: '#4a4a4a', backgroundColor: 'rgba(0,0,0,0)', marginLeft: 10}}>
                        {size == 'big' ? this.props.pickerStr : ''}
                    </Text>
                </View>
            )
        })
    }

    _onScrollEndDrag(e) {
        let y = e.nativeEvent.contentOffset.y;
        let onScrollEndDragCount = this.onScrollCount;
        let start = Date.now();
        if(this.fixInterval) {
            clearInterval(this.fixInterval);
        }
        this.fixInterval = setInterval(() => this._timeFix(start, y, onScrollEndDragCount),10);
    }

    _timeFix(start, y, onScrollEndDragCount) {
        let now = Date.now();
        let end = 200;
        if(now - start > end) {
            clearInterval(this.fixInterval);
            if(onScrollEndDragCount == this.onScrollCount) {
                this._onScrollEnd(y);
            }
        }
    }

    _onMomentumScrollEnd(e) {
        let y = e.nativeEvent.contentOffset.y;
        this._onScrollEnd(y);
    }

    _onScroll(e) {
        this.onScrollCount++;
        let y = e.nativeEvent.contentOffset.y;
        if(this.refs._ScrollView2) {
            this.refs._ScrollView2.scrollTo({y: y, animated: false});
        }
    }

    _onScrollEnd(y) {
        let y1 = y-(y%40);
        if(y%40 > 20) {y1 = y1+40;}
        let index = y1/40;
        if(this.refs._ScrollView) {
            this.refs._ScrollView.scrollTo({y: y1, animated: false});
        }
        if (this.props.onRowChange) {
            this.props.onRowChange(index);
        }
    }

    _selectTo = (index) => {
        let yy = index*40;
		const _this = this;
        if(this.refs._ScrollView) {
			var handle = findNodeHandle(this.refs._ScrollView);
			RCTUIManager.measure(handle, (x, y, width, height, pageX, pageY) => {
			  this.refs._ScrollView.scrollTo({x:0, y:yy, animated:true});
			});
        }
    }

    componentDidMount() {
        if(this.props.selectTo) {
            this._selectTo(this.props.selectTo);
        }
    }

    render() {
        return (
            <View style = {{flex: 1}}>
                <View style = {{height: 225, backgroundColor: '#ffffff'}}>
                    <ScrollView
                        bounces = {false}
                        onScrollEndDrag = {(e) => {this._onScrollEndDrag(e)}}
                        onMomentumScrollEnd = {(e) => {this._onMomentumScrollEnd(e)}}
                        onScroll = {(e) => {this._onScroll(e)}}
                        scrollEventThrottle = {16}
                        showsVerticalScrollIndicator = {false}
                        ref = '_ScrollView'
                    >
                        <View style = {{height: 100}} />
                        {this.getItem('small')}
                        <View style = {{height: 100}} />
                    </ScrollView>
                </View>
                <View style = {{height: 40, marginTop: -125, backgroundColor: '#ffffff'}} pointerEvents = 'none' >
                    <View style = {{height: CSS.pixel(1), backgroundColor: '#a2a2a2'}} />
                    <ScrollView
                        showsVerticalScrollIndicator = {false}
                        ref = '_ScrollView2'
                    >
                        {this.getItem('big')}
                    </ScrollView>
                    <View style = {{height: CSS.pixel(1), backgroundColor: '#a2a2a2'}} />
                </View>
                <View style = {{height: 100}}  pointerEvents = 'none' />
            </View>
        )
    }
}
